src/auth/mod.ts

```diff
@@ -15,33 +14,131 @@ interface DecodedToken {
 
 const logger = getLogger(import.meta);
 
-function isSubInPodsConfig(payload: DecodedToken): boolean {
-  return podsConfig.pods.some((pod) =>
-    pod.sub === payload.sub &&
-    pod.serviceaccount.uid === payload["kubernetes.io"]?.serviceaccount.uid
+export async function authenicateRequestAndReturnPodServiceAccount(
+  token: string,
+): Promise<string> {
+  const payload = await verifyToken(token);
+
+  if (!payload["kubernetes.io"]) {
+    logger.error("Payload does not contain kubernetes.io field");
+    throw new Error("Payload does not contain kubernetes.io field");
+  }
+
+  return payload["kubernetes.io"].serviceaccount.name;
+}
+
+async function verifyToken(token: string): Promise<DecodedToken> {
+  const key = await getPublicKey();
+
+  const cryptoKey = await crypto.subtle.importKey(
+    "jwk",
+    JSON.parse(key),
+    { name: "RSASSA-PKCS1-v1_5", hash: "SHA-256" },
+    true,
+    ["verify"],
   );
+  const verified = await verify(token, cryptoKey);
+  return verified as DecodedToken;
 }
 
-export function decodeToken(token: string): DecodedToken | null {
-  if (token.startsWith("Bearer ")) {
-    token = token.slice(7);
+async function getPublicKey(): Promise<string> {
+  const jwks_uri = await getJWKURI();
+  const jwkUrl = new URL(jwks_uri);
+  const headers = new Headers();
+  if (envVarsConfig.env === "DEV") {
+    // This should be the proxy server created by kubectl for dev purposes inside a local machine
+    jwkUrl.hostname = envVarsConfig.k8s_api;
+  } else {
+    // we need to set the auth headers with the serviceToken when herald is running in a k8s cluster
+    headers.set(
+      "Authorization",
+      `Bearer ${await getServiceAccountToken()}`,
+    );
+  }
+  const fetchJWK = await fetch(jwkUrl.toString(), { headers });
+  const data = await fetchJWK.json();
+  const key = data.keys[0].kid;
+  if (!key) {
+    throw new Error("Key not found in response");
   }
-  const [_headers, payload, _signature] = decode(token);
 
-  const data = payload as DecodedToken;
-  if (!data || typeof data !== "object" || !data.sub) {
-    logger.warn("Invalid token");
-    throw new HTTPException(401, {
-      message: "Invalid token",
-    });
+  return key;
+}
+
+async function getJWKURI(): Promise<string> {
+  const env = envVarsConfig.env;
```

Comment:
I am a bit lost between "envVarsConfig" and Deno.env... all the env should be parsed/used only in config.ts and then, only the config is used throughout the codebase

Reply:
I was refraining from adding those ones as they are specific to `DEV` env't